File: buffer.cpp
#include <string>
#include <vector>

class SlateBuffer
{
    public:
        SlateBuffer(int id, std::string title = "", std::string path = "");
    protected:
        std::string basename();

        int id;
        std::string title;
        std::string path;
        std::vector<std::string> lines;
};

/**
 * @param id    unique id for this buffer
 * @param title title for this buffer, empty if the buffer is for a file
 * @param path  path to file opened by this buffer, empty string if 
 *              the buffer is not for a file
 */
SlateBuffer::SlateBuffer(int id, std::string title, std::string path)
{
    this->id = id;

    if (title != "") {
        this->title = title;
    }

    if (path != "") {
        this->path = path;
    }
}

File: api/buffer.lua
---@class Buffers
---@field active_bufid integer
---@field open_bufs integer[]
---@field visible_bufs integer[]
local M = {}

-- TODO: maybe move buffer commands to separate module if slate module grows too big
---Vertically split the window
M.vsplit = function() end

---Horizontally split the window
M.hsplit = function() end

---@param bufid integer
M.set_active_buffer = function(bufid)
  M.active_bufid = bufid
end

---@param file_path? string path to file to open, opens empty scratch buffer if nil
M.open_buffer = function(file_path) end

---@param bufid integer id of buffer to close
M.close_buffer = function(bufid) end

---Begins visual selection at given cursor position
---@param cursor_x integer
---@param cursor_y integer
M.begin_selection = function(cursor_x, cursor_y) end

M.end_selection = function() end

return M
